### PR TITLE
Update TSM engine, WAL and encoding

### PR DESCRIPTION
**Pull request by pauldix**
* Add InfluxQLType to Values to map the TSM type to InfluxQL
* Fix bug in WAL where close wouldn't nil out the currentSegment after closing it
* Export writeSnapshot to be used in tests, add argument to run it async or not
* Update reloadCache to load temporary metadata information in the engine
* Update LoadMetadataIndex to use the temp WAL metadata information

I'm concerned about how this will impact startup performance, but we'll have to do some testing. Bunch of stuff happening in the WAL reload.

cc @jwilder @otoolep

### CODE REVIEW

File: tsdb/engine/tsm1/encoding.go [hunk 1]
```diff
@@ -100,6 +101,26 @@ func (a Values) Encode(buf []byte) ([]byte, error) {
 	return nil, fmt.Errorf("unsupported value type %T", a[0])
 }
 
+// InfluxQLType returns the influxql.DataType the values map to.
+func (a Values) InfluxQLType() (influxql.DataType, error) {
+	if len(a) == 0 {
+		return influxql.Unknown, fmt.Errorf("no values in collection")
```
Comment (otoolep) [inline tsdb/engine/tsm1/encoding.go]: Ruby lives! :-)

File: tsdb/engine/tsm1/engine.go [hunk 2]
```diff
@@ -149,27 +155,48 @@ func (e *DevEngine) LoadMetadataIndex(shard *tsdb.Shard, index *tsdb.DatabaseInd
...
 				return err
 			}
-		default:
-			return fmt.Errorf("unkown block type for: %v. got %v", k, typ)
 		}
+	}
+	e.walFields = nil
 
+	// load the series from the WAL
+	for seriesKey, measurement := range e.walSeries {
```
Comment (jwilder) [inline tsdb/engine/tsm1/engine.go]: Wonder if we could use the cache instead since it already has the same information.

File: tsdb/engine/tsm1/engine.go [hunk 3]
```diff
@@ -228,7 +258,8 @@ func (e *DevEngine) Begin(writable bool) (tsdb.Tx, error) {
 
 func (e *DevEngine) WriteTo(w io.Writer) (n int64, err error) { panic("not implemented") }
 
-func (e *DevEngine) writeSnapshot() error {
+// WriteSnapshot will snapshot the cache and write a new TSM file with its contents, releasing the snapshot when done.
+func (e *DevEngine) WriteSnapshot(async bool) error {
```
Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: What exactly does allowing this happen async get us?

Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: Compaction already runs in a goroutine, so why a second goroutine?

File: tsdb/engine/tsm1/engine.go [hunk 4]
```diff
@@ -357,6 +403,28 @@ func (e *DevEngine) reloadCache() error {
...
 }
 
+// loadWALMetadata will put the series key, measurement name, field name and type into the temporary wal maps of metadata
+func (e *DevEngine) loadWALMetadata(vals map[string][]Value) {
+	for k, v := range vals {
+		seriesKey, field := seriesAndFieldFromCompositeKey(k)
+		measurement := tsdb.MeasurementFromSeriesKey(seriesKey)
+		fieldType, err := Values(v).InfluxQLType()
+		if err != nil {
+			continue
```
Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: Should we log this?

File: tsdb/engine/tsm1/engine.go (hunk 3 above)
Comment (jwilder) [inline tsdb/engine/tsm1/engine.go]: Yeah, I'm not keen on this.  I would like see some data or benchmarks that can point to a problem that this is solving.  The first goroutine introduced a race condition after it was merged yesterday and I'm not sure what unintended issues this could introduce.

Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: Yeah, I get concerned when I see goroutines fired off without proper tracking of the routine. What if it takes longer to run that we expect, encounters some error that the `compact` routine should actually know about etc etc.

File: tsdb/engine/tsm1/engine.go (hunk 2 above)
Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: I thought of that, but it doesn't have the type information. To be honest, I don't think the type information is actually used in queries against TSM. I'm happy to remove this and iterate through only the Cache. Will be MUCH better.

File: tsdb/engine/tsm1/engine.go (hunk 3 above)
Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: Compaction runs in a single goroutine. In our testing with TSM1 we found that we became CPU bound by a single CPU (the one doing the compaction). Having this means that two snapshots can run at the same time on different CPUs.

@otoolep based on your comment we already have a problem with the `compact` goroutine not sending any information back to everything else.

In fact, this brings up a good point, which is that in many places we're logging errors and continuing on despite the fact that the errors should probably be irrecoverable (i.e. panic).

File: tsdb/engine/tsm1/engine.go (hunk 4 above)
Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: we should, will update

File: tsdb/engine/tsm1/engine.go (hunk 3 above)
Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: @pauldix -- I take your point about it's always awkward getting errors back from goroutines, but it was about limiting the exposure to this problem. Since the working with the cache snapshots is an integral part of the compaction process, it was at least all in the same goroutine. 

Does this system support multiple cache-snapshots being written to disk at once? If so, launching it from the compaction goroutine is unnecessary. If this snapshot process is truly decoupled from compaction, then we should run it in its own goroutine (or multiple goroutines if we want more work done in parallel). It would make the code clearer and easier to understand.

Comment (jwilder) [inline tsdb/engine/tsm1/engine.go]: @pauldix That is very different code then is what is here now.  I would like to see data that the current code exhibits the same issue.  If it is CPU bound, it's very likely we can tune it to reduce CPU.  The most CPU intensive part of writing a snapshot _should_ be the encoding which has been tuned quite a bit and should support a rate much faster than we're snapshotting.  If it's not the encoding, then we need to examine that first.  Throwing more goroutines at the problem is not right solution based on what I have seen.  In fact, it creates new problems with contention, potential race conditions as well as makes the code harder to reason about.

In regards to panics, the code should absolutely not be panicing in these cases.  We need to design a solution that handles errors appropriately.

Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: Multiple snapshots can be run at the same time. This is not currently true for compaction (and I'm not sure we ever need to worry about it for compaction).

Snapshots can quickly become a bottleneck because the WAL can stream data to disk very quickly. Having multiple snapshots run at the same time helps alleviate this.

I'm not sure what having a long running goroutine dedicated to snapshots gives us. How's that different than just kicking off goroutine as needed?

Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: I'm fine with ripping out the goroutine until we have more performance data. However, there are cases where I'd argue we should be panicing the server. Particularly in these early stages while we're just testing.

We don't want the server continually running and taking in writes despite the fact that things aren't working under the hood. A panic that sets off alarms is something that system operators can look into. Error messages that go into logs may or may not be checked on within any decent time frame.

We've seen this come up time and again over the last 2 years. People complained when the system would continue to take writes despite the fact that it was so overloaded that it wouldn't recover.

If your system runs out of HD space, for instance. A panic is appropriate. It will require a system operator to fix and you want to have alarms going off, which is something that a processing going down will usually trigger.

Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: >I'm not sure what having a long running goroutine dedicated to snapshots gives us. How's that different than just kicking off goroutine as needed?

From a black-box point of view, it doesn't make much difference. From an implementation point of view it is significant. Right now the goroutine to kick off snapshot-to-tsm is in the compaction goroutine. To the next guy looking at this code it's not clear whether this coupling is co-incidental, or is a key part of the design (it's the former it seems). Splitting it out into a separate goroutine -- which in turn can launch short-running goroutines if performance analysis shows us we should -- makes it crystal clear to the next guy how the components interact (or don't).

In other words it seems like unnecessary coupling of two distinct services in the engine.

Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: As for panicking on a full disk, can't we just do log.Fatal()? Does it have to panic?

Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: yes, `log.Fatal` would be better. But I think that system level errors from the file operations should exit the process instead of simply logging it and continuing to operate.

Comment (jwilder) [inline tsdb/engine/tsm1/engine.go]: There are multiple ways to handle all of those situations without crashing and still alert operators to problems that may need intervention.  If a database does panic, it should be for a very good reason.  Out of disk space is not a good reason IMO.  We should expect that to happen and design a solution that handles it gracefully.  Relying on your database crashing to alert you to the fact that you are not properly monitoring your disk usage does not make sense to me.

File: tsdb/engine/tsm1/engine.go (hunk 2 above)
Comment (jwilder) [inline tsdb/engine/tsm1/engine.go]: I think the type info might get used via Ben's query refactoring.  Not sure.  You can infer the the type from the value of the points for the key.  You're right that the current TSM code doesn't use it now though.

File: tsdb/engine/tsm1/engine.go (hunk 3 above)
Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: Here's another situation. Say snapshots fail to save for some underlying system reason. Could be file permissions, full HD, or whatever.

We continue to take writes until we fill up memory then reject them (because everything is backing up in the WAL). The user restarts the process, we take a long time to reload the cache (so now because we didn't error hard early, it takes a longer time to recover), and we immediately fill up the cache to the point where we can no longer accept writes again.

There are many situations in which it's more user friendly to let the process go down so the user can quickly look into it. I'm willing to be that there are many people out there that are not tailing their logs for errors. It's even worse because who knows how many errors we may be throwing out? We don't have levels clearly defined.

It's a poor user experience to throw up our hands and say "well you should have had better monitoring".

Continuing to answer queries, but throw writes on the floor because there are underlying system problems is a very poor experience. I still maintain that a system level error on writes should stop everything cold. Otherwise users will complain that we lost their data because we kept the database running (even if we returned 500s on the writes). And what about UDP? There's no way for the client to know.

**Comment by pauldix** [comment]
@jwilder, @otoolep just pushed up what I think addresses all comments

File: tsdb/engine/tsm1/engine.go [hunk 5]
```diff
@@ -127,58 +127,81 @@ func (e *DevEngine) Close() error {
...
-			if err := mf.CreateFieldIfNotExists(field, influxql.String, false); err != nil {
-				return err
-			}
-		default:
-			return fmt.Errorf("unkown block type for: %v. got %v", k, typ)
+		keysLoaded[k] = true
+	}
+
+	// load metadata from the Cache
+	e.Cache.mu.Lock() // shouldn't need the lock, but just to be safe
```
Comment (jwilder) [inline tsdb/engine/tsm1/engine.go]: The engine should not be locking the cache this way.  `mu` is not exported and should not be accessed outside of the `Cache`.  If you do need to expose locking the cache externally, I would do it via the `Locker` interface: https://golang.org/pkg/sync/#Locker 

I don't think the engine is actually fully opened when `LoadMetaDataIndex` is run so not sure that the lock is needed at all.

File: tsdb/engine/tsm1/engine.go [hunk 6]
```diff
@@ -127,58 +127,81 @@ func (e *DevEngine) Close() error {
...
+	defer e.Cache.mu.Unlock()
+
+	for key, entry := range e.Cache.store {
```
Comment (jwilder) [inline tsdb/engine/tsm1/engine.go]: `Store` should be exported to.

File: tsdb/engine/tsm1/engine.go (hunk 5 above)
Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: it isn't open at this point.

**Comment by jwilder** [comment]
Two small items but otherwise :+1:

File: tsdb/engine/tsm1/engine.go (hunk 6 above)
Comment (pauldix) [inline tsdb/engine/tsm1/engine.go]: I can update that, but it's tricky. Either I have to make a complete copy, or I make the `Cache` implement Locker as you mentioned. But then the store is accessible with no guarantees that it's safe.

Comment (jwilder) [inline tsdb/engine/tsm1/engine.go]: I'd go w/ `Locker` then.

**Comment by pauldix** [comment]
@jwilder done!

**Comment by jwilder** [comment]
:+1: on green

**Comment by otoolep** [comment]
+1
